besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java

```diff
@@ -558,7 +560,12 @@ public Runner build() {
   }
 
   private Optional<NatManager> buildNatManager(final NatMethod natMethod) {
-    switch (natMethod) {
+
+    final NatMethod detectedNatMethod =
+        Optional.of(natMethod)
+            .filter(not(isEqual(NatMethod.NONE)))
+            .orElseGet(NatService::autoDetectNatMethod);
```

Comment:
`orElseGet` takes a `Supplier`, which has 0 parameters. Woundn't calling `autoDetectNatMethod` with 0 arguments always result in `NatMethod.NONE`?

Reply:
Yes indeed. for now the scope is to set up an API to allow the detection of the NAT environment automatically. As for now there is no implementation which allows to automatically detect I used the `orElseGet`. There will be an implementation during the integration of docker and kubernetes (if it is possible to detect). For example, in the case of docker integration this code will change as below

```java
    final NatMethod detectedNatMethod =
        Optional.of(natMethod)
            .filter(not(isEqual(NatMethod.NONE)))
            .orElse(NatService.autoDetectNatMethod(
                  new DockerNatMethodAutoDetection(),  
                  new KubernetesNatMethodAutoDetection())
            );
    switch (detectedNatMethod) {
      case UPNP:
        return Optional.of(new UpnpNatManager());
      case MANUAL:
        return Optional.of(
            new ManualNatManager(p2pAdvertisedHost, p2pListenPort, jsonRpcConfiguration.getPort()));
      case NONE:
      default:
        return Optional.empty();
    }
```

Reply:
Ok, let's make the scope of the PRs a bit bigger so there's more context :)

Reply:
yes it's difficult to understand the context like that.  the integration of Docker and kubernetes is not within the scope of this bounty.  normally I have to work on it in a next bounty.  :)